massa-grpc/src/stream/send_blocks.rs

```diff
@@ -0,0 +1,212 @@
+// Copyright (c) 2023 MASSA LABS <info@massa.net>
+
+use crate::error::{match_for_io_error, GrpcError};
+use crate::server::MassaGrpc;
+use futures_util::StreamExt;
+use massa_models::block::{BlockDeserializer, BlockDeserializerArgs, SecureShareBlock};
+use massa_models::error::ModelsError;
+use massa_models::secure_share::SecureShareDeserializer;
+use massa_proto::google::rpc::Status;
+use massa_proto::massa::api::v1 as grpc;
+use massa_serialization::{DeserializeError, Deserializer};
+use std::io::ErrorKind;
+use std::pin::Pin;
+use tokio::sync::mpsc::Sender;
+use tonic::codegen::futures_core;
+use tonic::Request;
+use tracing::log::{error, warn};
+
+/// Type declaration for SendBlockStream
+pub type SendBlocksStream = Pin<
+    Box<
+        dyn futures_core::Stream<Item = Result<grpc::SendBlocksStreamResponse, tonic::Status>>
+            + Send
+            + 'static,
+    >,
+>;
+
+/// This function takes a streaming request of block messages,
+/// verifies, saves and propagates the block received in each message, and sends back a stream of
+/// block id messages
+pub(crate) async fn send_blocks(
+    grpc: &MassaGrpc,
+    request: Request<tonic::Streaming<grpc::SendBlocksStreamRequest>>,
+) -> Result<SendBlocksStream, GrpcError> {
+    let consensus_controller = grpc.consensus_controller.clone();
+    let mut protocol_command_sender = grpc.protocol_command_sender.clone();
+    let storage = grpc.storage.clone_without_refs();
+    let config = grpc.grpc_config.clone();
+
+    // Create a channel to handle communication with the client
+    let (tx, rx) = tokio::sync::mpsc::channel(config.max_channel_size);
+    // Extract the incoming stream of block messages
+    let mut in_stream = request.into_inner();
+
+    // Spawn a task that reads incoming messages and processes the block in each message
+    tokio::spawn(async move {
+        while let Some(result) = in_stream.next().await {
+            match result {
+                Ok(req_content) => {
+                    let Some(proto_block) = req_content.block else {
+                        report_error(
+                            req_content.id.clone(),
+                            tx.clone(),
+                            tonic::Code::InvalidArgument,
+                            "the request payload is empty".to_owned(),
+                        ).await;
+                        continue;
+                    };
+
+                    let args = BlockDeserializerArgs {
+                        thread_count: config.thread_count,
+                        max_operations_per_block: config.max_operations_per_block,
+                        endorsement_count: config.endorsement_count,
+                    };
+                    // Deserialize and verify received block in the incoming message
+                    match SecureShareDeserializer::new(BlockDeserializer::new(args))
+                        .deserialize::<DeserializeError>(&proto_block.serialized_data)
```

Comment:
The behavior here differ from the endorsements one. Here you use the `SecureShare` deserializer to deserialize the content but it should be used for the whole `SecureShare` object. Either you use the block deserializer directly or you deserialize the whole object as it's done in your `send_endorsements`

Reply:
TBH, I didn't understand what's wrong and what should I do instead